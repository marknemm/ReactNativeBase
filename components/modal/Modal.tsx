import Backdrop from '@components/backdrop/Backdrop';
import { Dialog } from '@rneui/themed';
import { View } from 'react-native';
import Animated, { FadeIn, FadeOut } from 'react-native-reanimated';
import type { ModalProps } from './Modal.interfaces';
import { useStyles } from './Modal.styles';

/**
 * Component for a blocking modal dialog.
 *
 * Unlike the standard {@link Dialog} component, the {@link Modal} component is blocking and is meant
 * to be launched imperatively from anywhere within the app via {@link import('@util/modal').showModal showModal}.
 *
 * @param props The component {@link ModalProps}.
 * @returns The {@link Modal} component.
 */
const Modal: React.FC<ModalProps> = ({
  backdropStyle,
  children,
  entering = FadeIn,
  exiting = FadeOut,
  isVisible = true,
  onBackdropPress,
  style,
}) => {
  const styles = useStyles({ style });

  return isVisible && (
    <>
      <Backdrop
        onPress={onBackdropPress}
        style={backdropStyle}
      />

      <Animated.View
        accessibilityViewIsModal
        entering={entering}
        exiting={exiting}
        pointerEvents="box-none"
        style={styles.modalContainer}
        testID="rnb-modal-container"
      >
        <View
          style={styles.modal}
          testID="rnb-modal"
        >
          {children}
        </View>
      </Animated.View>
    </>
  );
};

export type * from './Modal.interfaces';
export default Modal;
