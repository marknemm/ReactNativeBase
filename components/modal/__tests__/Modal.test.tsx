import Modal from '@components/modal/Modal';
import { Text } from '@rneui/themed';
import AppProvider from '@test/contexts/app/AppProvider';
import { fireEvent, render, screen } from '@testing-library/react-native';

describe('<Modal />', () => {
  const modalContentStr = 'Modal content';

  describe('isVisible prop', () => {
    it('should render the modal with children when isVisible is true (default)', () => {
      render(
        <Modal>
          <Text>{modalContentStr}</Text>
        </Modal>,
        { wrapper: AppProvider }
      );

      const modalText = screen.getByText(modalContentStr);
      expect(modalText).toBeVisible();
    });

    it('should not render the modal when isVisible is false', () => {
      render(
        <Modal isVisible={false}>
          <Text>{modalContentStr}</Text>
        </Modal>,
        { wrapper: AppProvider }
      );

      const modalText = screen.queryByText(modalContentStr);
      expect(modalText).toBeFalsy();
    });
  });

  describe('backdrop press', () => {
    it('should call onBackdropPress when backdrop is pressed', () => {
      const onBackdropPress = jest.fn();

      render(
        <Modal onBackdropPress={onBackdropPress}>
          <Text>{modalContentStr}</Text>
        </Modal>,
        { wrapper: AppProvider }
      );

      const backdrop = screen.getByTestId('rnb-backdrop');
      fireEvent(backdrop, 'touchEnd');

      expect(onBackdropPress).toHaveBeenCalledTimes(1);
    });

    it('should not call onBackdropPress when modal is pressed', async () => {
      const onBackdropPress = jest.fn();

      render(
        <Modal onBackdropPress={onBackdropPress}>
          <Text>{modalContentStr}</Text>
        </Modal>,
        { wrapper: AppProvider }
      );

      const modalText = screen.getByText(modalContentStr);
      fireEvent(modalText, 'touchEnd');

      expect(onBackdropPress).not.toHaveBeenCalled();
    });
  });

  describe('styles', () => {
    it('should apply the style prop to the modal', () => {
      const style = { backgroundColor: 'red' };

      render(
        <Modal style={style}>
          <Text>{modalContentStr}</Text>
        </Modal>,
        { wrapper: AppProvider }
      );

      const modal = screen.getByTestId('rnb-modal');
      expect(modal).toHaveStyle(style);
    });
  });

  describe('snapshots', () => {
    it('renders correctly when isVisible is true (default)', () => {
      render(
        <Modal />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly when isVisible is false', () => {
      render(
        <Modal isVisible={false} />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
