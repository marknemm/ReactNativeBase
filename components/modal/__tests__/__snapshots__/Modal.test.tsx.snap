// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> snapshots renders correctly when isVisible is false 1`] = `null`;

exports[`<Modal /> snapshots renders correctly when isVisible is true (default) 1`] = `
[
  <View
    collapsable={false}
    entering={[Function]}
    exiting={[Function]}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    style={
      {
        "backgroundColor": "#393e4233",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="rnb-backdrop"
  />,
  <View
    accessibilityViewIsModal={true}
    collapsable={false}
    entering={[Function]}
    exiting={[Function]}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    pointerEvents="box-none"
    style={
      {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="rnb-modal-container"
  >
    <View
      style={
        {
          "backgroundColor": "#ffffff",
          "borderRadius": 7,
          "height": 1020.1176470588235,
          "marginHorizontal": 34.285714285714285,
          "marginVertical": 156.94117647058823,
          "position": "absolute",
          "width": 681.4285714285714,
        }
      }
      testID="rnb-modal"
    />
  </View>,
]
`;
